### PR TITLE
add support for cosmosdb sql database

### LINKED ISSUE #3181: Add support for azure.cosmos.database.v0 resource type
[This PR in the Aspire repo](https://github.com/dotnet/aspire/pull/1583) is adding a new resource type for Aspire called `azure.cosmosdb.database.v0`.

We need this for Cosmos because the access token that the client code gets in an Aspire application is not scoped for database creation. You need to get a service management token to do that and use the `Azure.ResourceManager` APIs to create the database. For other resource types such as SQL Server and Postgres we are going to be creating the database instances as well so this just aligns the capabilities.

### PR DESCRIPTION
**Pull request by hemarina**
fix #3181
add cosmos db sql db

### CODE REVIEW

**Comment by mitchdenny** [comment]
I'll try this out today and confirm we are unblocked

File: cli/azd/pkg/infra/azure_resource_types.go [hunk 1]
```diff
@@ -14,6 +14,7 @@ const (
 	AzureResourceTypeCacheForRedis           AzureResourceType = "Microsoft.Cache/redis"
 	AzureResourceTypeCDNProfile              AzureResourceType = "Microsoft.Cdn/profiles"
 	AzureResourceTypeCosmosDb                AzureResourceType = "Microsoft.DocumentDB/databaseAccounts"
+	AzureResourceTypeCosmosDbSqlDb           AzureResourceType = "Microsoft.DocumentDB/databaseAccounts/sqlDatabases"
```
[the remark this answers is not in the document]
Comment (vhvb1989) [reply cli/azd/pkg/infra/azure_resource_types.go:17]: IMO, we dont need to print it.
azd prints only the services, and for a Cosmos DB, it should be fine to only list the cosmos service.  Folks can go to the Azure Portal to see all the details about it.

File: cli/azd/pkg/infra/azure_resource_types.go [hunk 2]
```diff
@@ -80,6 +81,8 @@ func GetResourceTypeDisplayName(resourceType AzureResourceType) string {
 		return "App Service plan"
 	case AzureResourceTypeCosmosDb:
 		return "Azure Cosmos DB"
+	case AzureResourceTypeCosmosDbSqlDb:
+		return "Azure Cosmos DB SQL DB"
```
Comment (vhvb1989) [inline cli/azd/pkg/infra/azure_resource_types.go:85]: I don't think we want this.

File: cli/azd/resources/apphost/templates/resources.bicept [hunk 3]
```diff
@@ -309,10 +309,12 @@ resource {{bicepName $name}}RoleAssignment 'Microsoft.Authorization/roleAssignme
   }
 }
 {{- end}}
+
 {{range $name, $value := .CosmosDbAccounts}}
-resource {{bicepName $name}} 'Microsoft.DocumentDB/databaseAccounts@2023-11-15' = {
+resource {{bicepName $name}} 'Microsoft.DocumentDB/databaseAccounts@2023-04-15' = {
   name: replace('{{$name}}-${resourceToken}', '-', '')
   location: location
+  kind: 'GlobalDocumentDB'
```
Comment (vhvb1989) [inline cli/azd/resources/apphost/templates/resources.bicept:317]: why are we adding this?

File: cli/azd/resources/apphost/templates/resources.bicept [hunk 4]
```diff
@@ -309,10 +309,12 @@ resource {{bicepName $name}}RoleAssignment 'Microsoft.Authorization/roleAssignme
   }
 }
 {{- end}}
+
 {{range $name, $value := .CosmosDbAccounts}}
-resource {{bicepName $name}} 'Microsoft.DocumentDB/databaseAccounts@2023-11-15' = {
+resource {{bicepName $name}} 'Microsoft.DocumentDB/databaseAccounts@2023-04-15' = {
```
Comment (vhvb1989) [inline cli/azd/resources/apphost/templates/resources.bicept:314]: is there a reason to use this older version?

File: cli/azd/resources/apphost/templates/resources.bicept (hunk 3 above)
Comment (vhvb1989) [reply cli/azd/resources/apphost/templates/resources.bicept:317]: the Aspire provisioner ( https://github.com/dotnet/aspire/blob/main/src/Aspire.Hosting.Azure.Provisioning/Provisioners/AzureCosmosDBProvisioner.cs#L66) is not setting this explicitly. I'm guessing that might be the default `kind` .

I would still recommend checking with @mitchdenny about this. Azd must produce the same result as the aspire provisioner.

File: cli/azd/resources/apphost/templates/resources.bicept [hunk 5]
```diff
@@ -376,6 +436,6 @@ output SERVICE_BINDING_{{alphaSnakeUpper $cname}}_ENDPOINT string = {{bicepName
 output SERVICE_BINDING_{{alphaSnakeUpper $name}}_ENDPOINT string = {{bicepName $name}}.properties.vaultUri
 {{end -}}
 {{range $name, $value := .CosmosDbAccounts -}}
-output SERVICE_BINDING_{{alphaSnakeUpper $name}}_ENDPOINT string = {{bicepName $name}}.properties.documentEndpoint
+output connectionStringKey_{{alphaSnakeUpper $name}} string = {{bicepName $name}}.name
```
Comment (vhvb1989) [inline cli/azd/resources/apphost/templates/resources.bicept:439]: How is the name of the cosmos account the connection string?  I am confused here.

File: cli/azd/resources/apphost/templates/resources.bicept (hunk 3 above)
Comment (davidfowl) [reply cli/azd/resources/apphost/templates/resources.bicept:316]: cc @Pilchie

File: cli/azd/resources/apphost/templates/resources.bicept [hunk 6]
```diff
@@ -345,7 +346,95 @@ resource {{bicepName $name}}OperatorRoleAssignment 'Microsoft.Authorization/role
...
+    properties: {
+      resource: {
+        id: 'container-{{$cname}}'
+        partitionKey: {
+          paths: [
+          '/myPartitionKey'
+        ]
+          kind: 'Hash'
+        }
+      }
```
Comment (davidfowl) [inline cli/azd/resources/apphost/templates/resources.bicept:423]: Is this a good default?

Comment (mitchdenny) [reply cli/azd/resources/apphost/templates/resources.bicept:423]: Not sure why we are even creating a container?

**Comment by mitchdenny** [comment]
@hemarina we've had a conversation on the Aspire side, and we aren't entirely happy with having to spin up a KeyVault resource just for the Cosmos connection strings. If we are going to make a KeyVault resource we'd like it to be more obvious in our app model.

For preview 3, we want to just have the connection string injected into the environment variables. It isn't the final solution, but it eliminates the need for KeyVault until we figure out the way we want to go with parameter/secret handling int he app model.

Sorry for the extra work :|

**Comment by vhvb1989** [comment]
@mitchdenny , without a KeyVault, azd will need to pull the connectionString from the cosmos account during runtime (similar to how azd is pulling the connection strings from containerApps service).
This is because we can't just set the connectionString (including a key) as bicep output as that's a security warning.

**Comment by davidfowl** [comment]
> This is because we can't just set the connectionString (including a key) as bicep output as that's a security warning.

Interesting...

**Comment by vhvb1989** [comment]
@mitchdenny @davidfowl , I created this PR https://github.com/Azure/azure-dev/pull/3213 on top of @hemarina 's changes.

It is an alternative where we remove the Key Vault and pull the connectionstring directly from the cosmos account.  Is that how you were thinking?  Note that, using the connectionString (which has the primary key) will ignore RBAC, so we might not need to set roles anymore (if we want this strategy). We only need to make sure that the user can access the keys/connectionString from the created cosmos account.

**Comment by davidfowl** [comment]
I think we go with connection strings for preview3 and move to Managed idenity + RBAC in preview4.

**Comment by mitchdenny** [comment]
> @mitchdenny @davidfowl , I created this PR #3213 on top of @hemarina 's changes.
> 
> It is an alternative where we remove the Key Vault and pull the connectionstring directly from the cosmos account. Is that how you were thinking? Note that, using the connectionString (which has the primary key) will ignore RBAC, so we might not need to set roles anymore (if we want this strategy). We only need to make sure that the user can access the keys/connectionString from the created cosmos account.

Yes ... I was thinking that there would be a Bicep expression on the container app for an environment variable which references the Cosmos instance and uses `listConnectionStrings()`.

**Comment by davidfowl** [comment]
[This PR](https://github.com/Azure/azure-dev/pull/3213) does the needful 😄

**Comment by vhvb1989** [comment]
> > @mitchdenny @davidfowl , I created this PR #3213 on top of @hemarina 's changes.
> > It is an alternative where we remove the Key Vault and pull the connectionstring directly from the cosmos account. Is that how you were thinking? Note that, using the connectionString (which has the primary key) will ignore RBAC, so we might not need to set roles anymore (if we want this strategy). We only need to make sure that the user can access the keys/connectionString from the created cosmos account.
> 
> Yes ... I was thinking that there would be a Bicep expression on the container app for an environment variable which references the Cosmos instance and uses `listConnectionStrings()`.

We can't ref `listConnectionStrings()` on Bicep for the containerApp (the one hosting the aspire app) because it is not part of the bicep.
We only describe the Container Registry with bicep.
Azd creates a `manifest template` for each aspire-app that is used during deploying the app (after provisioning infra).  
On https://github.com/Azure/azure-dev/pull/3213, azd is adding `{{connectionString "ENV_NAME_WITH_COSMOS_ACCOUNT"}}` to the `manifest template` as a placeholder. During deploy, azd will first get the value from the .env for what is the cosmos account name (which is only known b/c it's a bicep output). Then it uses arm client to pull the connection string and replace it for the containerApp creation (in the payload) :)

There were a few test failing for the PR, but I think I have fixed it all and CI should produce an azd version from https://github.com/Azure/azure-dev/pull/3213 at any time now. Please use it for your tests.

This strategy might be applied for Azure SQL

File: cli/azd/resources/apphost/templates/resources.bicept (hunk 3 above)
Comment (Pilchie) [reply cli/azd/resources/apphost/templates/resources.bicept:316]: My team doesn't own the management SDK, but I expect that it is the default.  This question talks about the values a bit: https://stackoverflow.com/questions/70613680/how-does-azure-cosmos-db-apis-maps-in-cli-and-azure-portal

**Comment by ellismg** [comment]
@mitchdenny @davidfowl 

> we've had a conversation on the Aspire side, and we aren't entirely happy with having to spin up a KeyVault resource just for the Cosmos connection strings. If we are going to make a KeyVault resource we'd like it to be more obvious in our app model.

Why do you feel this way?  Given the fact that there isn't end-to-end support for managed identity/RBAC for all scenarios across Azure today, it feels like for a while we are going to be stuck with "you need this connection string which itself contains a password and should be stored security" and that Key Vault is the answer for that.  That's a problem that is not specific to Cosmos (e.g. some of the discussions about the "connection string" for a storage account has also touched on this).

What makes it okay for the Container Registry to not be uttered it the manifest (and we create a container registry on behalf of the user) but using KeyVault as an implementation detail in this way for flowing strings that have secrets in them unacceptable or unpalatable?

**Comment by vhvb1989** [comment]
If going for the Key Vault approach, this PR would still need some changes.

There should ever be *one only Key Vault* used by azd, no matter how many cosmos accounts or dbs are created.
The one only Key Vault should host all the secrets for cosmos, AzureSQL or any other service in the future (like Search or AI service).  

The implementation on this PR is proposing one Key Vault for every cosmos account.. which is not OK

**Comment by davidfowl** [comment]
> What makes it okay for the Container Registry to not be uttered it the manifest (and we create a container registry on behalf of the user) but using KeyVault as an implementation detail in this way for flowing strings that have secrets in them unacceptable or unpalatable?

I don't know that it is OK, but I feel like we need to spend more time thinking about what happens when the user then uses KeyVault in their model and they end up with 2.

**Comment by vhvb1989** [comment]
I will update https://github.com/Azure/azure-dev/pull/3213 and try using containerApp secrets to hold the connection string

**Comment by davidfowl** [comment]
> I will update https://github.com/Azure/azure-dev/pull/3213 and try using containerApp secrets to hold the connection string

This sounds much better!

**Pull request closed without merging**